tests/0136-resolve_cb.c

```diff
@@ -0,0 +1,176 @@
+ * connect callback is not invoked with the output of the resolution callback,
+ * the test will fail.
+ */
+
+/**
+ * Stage of the test:
+ *   0: expecting resolve_cb to be invoked with TESTING_RESOLVE_CB:1234
+ *   1: expecting resolve_cb to be invoked with NULL, NULL
+ *   2: expecting connect_cb to invoked with socket address 127.1.2.3:57616
+ *   3: done
+ */
+static rd_atomic32_t stage;
+
+/** Exposes current test struct (in TLS) to callbacks. */
+static struct test *this_test;
+
+static int resolve_cb(const char *node,
+                      const char *service,
+                      const struct addrinfo *hints,
+                      struct addrinfo **res,
+                      void *opaque) {
+
+        int32_t cnt;
+
+        test_curr = this_test;
+
+        TEST_SAY("resolve_cb invoked: node=%s service=%s\n", node, service);
+
+        cnt = rd_atomic32_get(&stage);
+
+        if (cnt == 0) {
+                /* Stage 0: return a bogus address. */
+
+                struct sockaddr_in *addr;
+
+                TEST_ASSERT(strcmp(node, "TESTING_RESOLVE_CB") == 0,
+                            "unexpected node: %s\n", node);
+                TEST_ASSERT(strcmp(service, "1234") == 0,
+                            "unexpected service: %s\n", service);
+
+                addr                  = calloc(1, sizeof(struct sockaddr_in));
+                addr->sin_family      = AF_INET;
+                addr->sin_port        = htons(4321);
+                addr->sin_addr.s_addr = htonl(0x7f010203) /* 127.1.2.3 */;
+
+                *res                = calloc(1, sizeof(struct addrinfo));
+                (*res)->ai_family   = AF_INET;
+                (*res)->ai_socktype = SOCK_STREAM;
+                (*res)->ai_protocol = IPPROTO_TCP;
+                (*res)->ai_addrlen  = sizeof(struct sockaddr_in);
+                (*res)->ai_addr     = (struct sockaddr *)addr;
+        } else if (cnt == 1) {
+                /* Stage 1: free the bogus address returned in stage 0. */
+
+                TEST_ASSERT(node == NULL);
+                TEST_ASSERT(service == NULL);
+                TEST_ASSERT(hints == NULL);
+                free((*res)->ai_addr);
+                free(*res);
+        } else {
+                /* Stage 2+: irrelevant, simply fail to resolve. */
+
+                return -1;
+        }
+
+        rd_atomic32_add(&stage, 1);
+        return 0;
+}
+
+static int connect_cb(int s,
+                      const struct sockaddr *addr,
+                      int addrlen,
+                      const char *id,
+                      void *opaque) {
+        /* Stage 3: assert address is expected bogus. */
+
+        int32_t cnt;
+        struct sockaddr_in *addr_in;
+
+        test_curr = this_test;
+
+        TEST_SAY("connect_cb invoked\n");
+
+        cnt = rd_atomic32_get(&stage);
+        TEST_ASSERT(cnt == 2, "connect_cb invoked in unexpected stage: %d\n",
```

Comment:
nit: TEST_ASSERT and TEST_FAIL should not have newlines

Reply:
Done.